File: src/core/agent/context/storage/database/reader.ts
import { CreateDatabaseClient } from './client';
import { DatabaseConfig } from './types';

/**
 * 作用：从数据库查询所有会话信息
 * 关联：被用户调用，用于获取所有会话列表
 * 预期结果：返回所有session的基本信息数组（不包含entries和compacted_entries）
 */
export async function GetAllSessionsFromDatabase(config: DatabaseConfig): Promise<any[]> {
    if (!config) {
        throw new Error('databaseConfig is required');
    }
    const db = CreateDatabaseClient(config);
    try {
        const result = await db.query(
            `SELECT session_id, status, create_at FROM veloca_session WHERE status = 0 ORDER BY create_at DESC`
        );
        return result.rows.map((row: any) => ({
            session_id: row.session_id,
            status: row.status,
            create_at: row.create_at
        }));
    } finally {
        await db.end();
    }
}

/**
 * 作用：根据session_id从数据库读取该会话的所有数据
 * 关联：被combineContext调用，读取指定会话的session、entries和compacted_entries
 * 预期结果：返回包含该会话所有相关数据的对象，如果session不存在则返回空数据结构
 */
export async function ReadSessionDataFromDatabase(
    sessionId: string,
    config: DatabaseConfig
): Promise<any> {
    if (!sessionId) {
        throw new Error('sessionId is required');
    }
    if (!config) {
        throw new Error('databaseConfig is required');
    }
    const db = CreateDatabaseClient(config);
    try {
        const sessionResult = await db.query(
            `SELECT session_id, status, create_at FROM veloca_session WHERE session_id = $1 AND status = 0`,
            [sessionId]
        );
        if (sessionResult.rows.length === 0) {
            return {
                session: null,
                entries: [],
                compacted_entries: []
            };
        }
        const session = {
            session_id: sessionResult.rows[0].session_id,
            status: sessionResult.rows[0].status,
            create_at: sessionResult.rows[0].create_at
        };
        const entriesResult = await db.query(
            `SELECT entry_id, session_id, content, role, token_consumption, status, tools, create_at, is_compaction
             FROM veloca_entries WHERE session_id = $1 AND status = 0 ORDER BY create_at ASC`,
            [sessionId]
        );
        const entries = entriesResult.rows.map((row: any) => ({
            entry_id: row.entry_id,
            session_id: row.session_id,
            content: row.content,
            role: row.role,
            token_consumption: row.token_consumption,
            status: row.status,
            tools: row.tools ? JSON.parse(row.tools) : null,
            create_at: row.create_at,
            is_compaction: row.is_compaction
        }));
        const compactedResult = await db.query(
            `SELECT entry_id, session_id, trigger_entry_id, summary, create_at, status
             FROM veloca_compacted_entries WHERE session_id = $1 AND status = 0 ORDER BY create_at ASC`,
            [sessionId]
        );
        const compacted_entries = compactedResult.rows.map((row: any) => ({
            entry_id: row.entry_id,
            session_id: row.session_id,
            trigger_entry_id: row.trigger_entry_id,
            summary: row.summary,
            create_at: row.create_at,
            status: row.status
        }));
        return {
            session,
            entries,
            compacted_entries
        };
    } finally {
        await db.end();
    }
}

File: crates/otherone-storage/src/database/reader.rs
use crate::error::StorageError;
use crate::types::{
    AttributeBag, CompactedEntry, DatabaseConfig, Entry, RuntimeContext, Session, SessionData,
};

use super::client::create_database_client;

pub async fn get_all_sessions_from_database(
    config: &DatabaseConfig,
) -> Result<Vec<Session>, StorageError> {
    get_all_sessions_from_database_with_context(config, &RuntimeContext::legacy_default()).await
}

pub async fn get_all_sessions_from_database_with_context(
    config: &DatabaseConfig,
    runtime_context: &RuntimeContext,
) -> Result<Vec<Session>, StorageError> {
    runtime_context
        .validate()
        .map_err(StorageError::ConfigError)?;

    let pool = create_database_client(config).await?;

    let rows = sqlx::query_as::<
        _,
        (
            String,
            String,
            i16,
            chrono::NaiveDateTime,
            Option<String>,
            Option<String>,
        ),
    >(
        "SELECT partition_key, session_id, status, create_at, attributes_json, metadata_json \
         FROM otherone_session \
         WHERE partition_key = $1 AND status = 0 \
         ORDER BY create_at DESC",
    )
    .bind(&runtime_context.partition_key)
    .fetch_all(&pool)
    .await?;

    let sessions = rows
        .into_iter()
        .map(
            |(partition_key, session_id, status, create_at, attributes_json, metadata_json)| {
                Session {
                    partition_key: Some(partition_key),
                    session_id,
                    status,
                    create_at: create_at.to_string(),
                    attributes: parse_attributes(attributes_json),
                    metadata: parse_attributes(metadata_json),
                }
            },
        )
        .collect();

    pool.close().await;
    Ok(sessions)
}

pub async fn read_session_data_from_database(
    session_id: &str,
    config: &DatabaseConfig,
) -> Result<SessionData, StorageError> {
    read_session_data_from_database_with_context(
        session_id,
        config,
        &RuntimeContext::legacy_default(),
    )
    .await
}

pub async fn read_session_data_from_database_with_context(
    session_id: &str,
    config: &DatabaseConfig,
    runtime_context: &RuntimeContext,
) -> Result<SessionData, StorageError> {
    runtime_context
        .validate()
        .map_err(StorageError::ConfigError)?;
    if session_id.is_empty() {
        return Err(StorageError::ConfigError(
            "session_id is required".to_string(),
        ));
    }

    let pool = create_database_client(config).await?;

    let session_result = sqlx::query_as::<
        _,
        (
            String,
            String,
            i16,
            chrono::NaiveDateTime,
            Option<String>,
            Option<String>,
        ),
    >(
        "SELECT partition_key, session_id, status, create_at, attributes_json, metadata_json \
         FROM otherone_session \
         WHERE partition_key = $1 AND session_id = $2 AND status = 0",
    )
    .bind(&runtime_context.partition_key)
    .bind(session_id)
    .fetch_optional(&pool)
    .await?;

    let session = match session_result {
        None => {
            pool.close().await;
            return Ok(SessionData {
                session: None,
                entries: Vec::new(),
                compacted_entries: Vec::new(),
            });
        }
        Some((partition_key, sid, status, create_at, attributes_json, metadata_json)) => Session {
            partition_key: Some(partition_key),
            session_id: sid,
            status,
            create_at: create_at.to_string(),
            attributes: parse_attributes(attributes_json),
            metadata: parse_attributes(metadata_json),
        },
    };

    let entries_rows = sqlx::query_as::<_, (String, String, String, String, String, Option<i32>, i16, Option<String>, chrono::NaiveDateTime, i16, Option<String>, Option<String>)>(
        "SELECT partition_key, entry_id, session_id, content, role, token_consumption, status, tools, create_at, is_compaction, attributes_json, metadata_json \
         FROM otherone_entries \
         WHERE partition_key = $1 AND session_id = $2 AND status = 0 \
         ORDER BY create_at ASC",
    )
    .bind(&runtime_context.partition_key)
    .bind(session_id)
    .fetch_all(&pool)
    .await?;

    let entries = entries_rows
        .into_iter()
        .map(
            |(
                partition_key,
                entry_id,
                sid,
                content,
                role,
                token_consumption,
                status,
                tools,
                create_at,
                is_compaction,
                attributes_json,
                metadata_json,
            )| Entry {
                partition_key: Some(partition_key),
                entry_id,
                session_id: sid,
                content,
                role,
                token_consumption: token_consumption.map(|value| value as u32),
                status,
                tools: tools.and_then(|value| serde_json::from_str(&value).ok()),
                create_at: create_at.to_string(),
                is_compaction,
                attributes: parse_attributes(attributes_json),
                metadata: parse_attributes(metadata_json),
            },
        )
        .collect();

    let compacted_rows = sqlx::query_as::<_, (String, String, String, String, String, chrono::NaiveDateTime, i16, Option<String>, Option<String>)>(
        "SELECT partition_key, entry_id, session_id, trigger_entry_id, summary, create_at, status, attributes_json, metadata_json \
         FROM otherone_compacted_entries \
         WHERE partition_key = $1 AND session_id = $2 AND status = 0 \
         ORDER BY create_at ASC",
    )
    .bind(&runtime_context.partition_key)
    .bind(session_id)
    .fetch_all(&pool)
    .await?;

    let compacted_entries = compacted_rows
        .into_iter()
        .map(
            |(
                partition_key,
                entry_id,
                sid,
                trigger_entry_id,
                summary,
                create_at,
                status,
                attributes_json,
                metadata_json,
            )| CompactedEntry {
                partition_key: Some(partition_key),
                entry_id,
                session_id: sid,
                trigger_entry_id,
                summary,
                create_at: create_at.to_string(),
                status,
                attributes: parse_attributes(attributes_json),
                metadata: parse_attributes(metadata_json),
            },
        )
        .collect();

    pool.close().await;

    Ok(SessionData {
        session: Some(session),
        entries,
        compacted_entries,
    })
}

fn parse_attributes(raw: Option<String>) -> AttributeBag {
    raw.and_then(|value| serde_json::from_str::<AttributeBag>(&value).ok())
        .unwrap_or_default()
}
